## Importing Libraries 

In [1]:
import pandas as pd
import numpy as np
import os
import sqlite3
import seaborn as sns

In [2]:
# Get the current working directory
cwd = os.getcwd()

# Print the current working directory
print("Current working directory: {0}".format(cwd))

Current working directory: C:\Users\geral\Repository\Coursework


In [3]:
# Change the current working directory
os.chdir(r'C:\Users\geral\Repository\Coursework\Dataset')

# Print the current working directory
print("Current working directory: {0}".format(os.getcwd()))

Current working directory: C:\Users\geral\Repository\Coursework\Dataset


## Connecting to database

In [4]:
conn = sqlite3.connect('pydatabase.db')

In [5]:
c = conn.cursor()

### Importing dataset

In [5]:
airports = pd.read_csv('airports.csv')

In [6]:
airports.head()

iata               airport              city state country        lat  \
0  00M              Thigpen        Bay Springs    MS     USA  31.953765   
1  00R  Livingston Municipal        Livingston    TX     USA  30.685861   
2  00V           Meadow Lake  Colorado Springs    CO     USA  38.945749   
3  01G          Perry-Warsaw             Perry    NY     USA  42.741347   
4  01J      Hilliard Airpark          Hilliard    FL     USA  30.688012   

         long  
0  -89.234505  
1  -95.017928  
2 -104.569893  
3  -78.052081  
4  -81.905944

In [7]:
airports.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3376 entries, 0 to 3375
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   iata     3376 non-null   object 
 1   airport  3376 non-null   object 
 2   city     3364 non-null   object 
 3   state    3364 non-null   object 
 4   country  3376 non-null   object 
 5   lat      3376 non-null   float64
 6   long     3376 non-null   float64
dtypes: float64(2), object(5)
memory usage: 184.8+ KB


In [8]:
carriers = pd.read_csv('carriers.csv')

In [9]:
carriers.head()

Code                    Description
0  02Q                  Titan Airways
1  04Q             Tradewind Aviation
2  05Q            Comlux Aviation, AG
3  06Q  Master Top Linhas Aereas Ltd.
4  07Q            Flair Airlines Ltd.

In [10]:
carriers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1491 entries, 0 to 1490
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Code         1490 non-null   object
 1   Description  1491 non-null   object
dtypes: object(2)
memory usage: 23.4+ KB


In [11]:
plane_data = pd.read_csv('plane_data.csv')

In [12]:
plane_data.head()

tailnum type manufacturer issue_date model status aircraft_type engine_type  \
0  N050AA  NaN          NaN        NaN   NaN    NaN           NaN         NaN   
1  N051AA  NaN          NaN        NaN   NaN    NaN           NaN         NaN   
2  N052AA  NaN          NaN        NaN   NaN    NaN           NaN         NaN   
3  N054AA  NaN          NaN        NaN   NaN    NaN           NaN         NaN   
4  N055AA  NaN          NaN        NaN   NaN    NaN           NaN         NaN   

  year  
0  NaN  
1  NaN  
2  NaN  
3  NaN  
4  NaN

In [13]:
plane_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5029 entries, 0 to 5028
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   tailnum        5029 non-null   object
 1   type           4480 non-null   object
 2   manufacturer   4480 non-null   object
 3   issue_date     4480 non-null   object
 4   model          4480 non-null   object
 5   status         4480 non-null   object
 6   aircraft_type  4480 non-null   object
 7   engine_type    4480 non-null   object
 8   year           4480 non-null   object
dtypes: object(9)
memory usage: 353.7+ KB


In [12]:
Y2005 = pd.read_csv('2005.csv')

In [39]:
Y2005.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2005      1          28          5   1603.0        1605   1741.0   
1  2005      1          29          6   1559.0        1605   1736.0   
2  2005      1          30          7   1603.0        1610   1741.0   
3  2005      1          31          1   1556.0        1605   1726.0   
4  2005      1           2          7   1934.0        1900   2235.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        1759            UA        541  ...      4       23          0   
1        1759            UA        541  ...      6       15          0   
2        1805            UA        541  ...      9       18          0   
3        1759            UA        541  ...     11       10          0   
4        2232            UA        542  ...      5       10          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0               NaN         0             0            0        0   
1               NaN         0             0            0        0   
2               NaN         0             0            0        0   
3               NaN         0             0            0        0   
4               NaN         0             0            0        0   

   SecurityDelay  LateAircraftDelay  
0              0                  0  
1              0                  0  
2              0                  0  
3              0                  0  
4              0                  0  

[5 rows x 29 columns]

In [7]:
Y2005.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7140596 entries, 0 to 7140595
Data columns (total 29 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Year               int64  
 1   Month              int64  
 2   DayofMonth         int64  
 3   DayOfWeek          int64  
 4   DepTime            float64
 5   CRSDepTime         int64  
 6   ArrTime            float64
 7   CRSArrTime         int64  
 8   UniqueCarrier      object 
 9   FlightNum          int64  
 10  TailNum            object 
 11  ActualElapsedTime  float64
 12  CRSElapsedTime     int64  
 13  AirTime            float64
 14  ArrDelay           float64
 15  DepDelay           float64
 16  Origin             object 
 17  Dest               object 
 18  Distance           int64  
 19  TaxiIn             int64  
 20  TaxiOut            int64  
 21  Cancelled          int64  
 22  CancellationCode   object 
 23  Diverted           int64  
 24  CarrierDelay       int64  
 25  WeatherDelay      

In [40]:
Y2006 = pd.read_csv('2006.csv')

In [41]:
Y2006.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2006      1          11          3    743.0         745   1024.0   
1  2006      1          11          3   1053.0        1053   1313.0   
2  2006      1          11          3   1915.0        1915   2110.0   
3  2006      1          11          3   1753.0        1755   1925.0   
4  2006      1          11          3    824.0         832   1015.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        1018            US        343  ...     45       13          0   
1        1318            US        613  ...     27       19          0   
2        2133            US        617  ...      4       11          0   
3        1933            US        300  ...     16       10          0   
4        1015            US        765  ...     27       12          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0               NaN         0             0            0        0   
1               NaN         0             0            0        0   
2               NaN         0             0            0        0   
3               NaN         0             0            0        0   
4               NaN         0             0            0        0   

   SecurityDelay  LateAircraftDelay  
0              0                  0  
1              0                  0  
2              0                  0  
3              0                  0  
4              0                  0  

[5 rows x 29 columns]

In [9]:
Y2006.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7141922 entries, 0 to 7141921
Data columns (total 29 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Year               int64  
 1   Month              int64  
 2   DayofMonth         int64  
 3   DayOfWeek          int64  
 4   DepTime            float64
 5   CRSDepTime         int64  
 6   ArrTime            float64
 7   CRSArrTime         int64  
 8   UniqueCarrier      object 
 9   FlightNum          int64  
 10  TailNum            object 
 11  ActualElapsedTime  float64
 12  CRSElapsedTime     float64
 13  AirTime            float64
 14  ArrDelay           float64
 15  DepDelay           float64
 16  Origin             object 
 17  Dest               object 
 18  Distance           int64  
 19  TaxiIn             int64  
 20  TaxiOut            int64  
 21  Cancelled          int64  
 22  CancellationCode   object 
 23  Diverted           int64  
 24  CarrierDelay       int64  
 25  WeatherDelay      

## Uploading into tables on the database 

In [14]:
airports.to_sql('airports', con = conn, index = False)

In [15]:
carriers.to_sql('carriers', con = conn, index = False)

In [16]:
plane_data.to_sql('plane_data', con = conn, index = False)

In [26]:
Y2005.to_sql('2005', con = conn, index = False)

In [27]:
Y2006.to_sql('2006', con = conn, index = False)

Combining Y2005 and Y2006 to Y05_06, then uploading on to the database

In [42]:
Y05_06 = Y2005.append(other = Y2006)

In [10]:
Y05_06.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2005      1          28          5   1603.0        1605   1741.0   
1  2005      1          29          6   1559.0        1605   1736.0   
2  2005      1          30          7   1603.0        1610   1741.0   
3  2005      1          31          1   1556.0        1605   1726.0   
4  2005      1           2          7   1934.0        1900   2235.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        1759            UA        541  ...      4       23          0   
1        1759            UA        541  ...      6       15          0   
2        1805            UA        541  ...      9       18          0   
3        1759            UA        541  ...     11       10          0   
4        2232            UA        542  ...      5       10          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0               NaN         0             0            0        0   
1               NaN         0             0            0        0   
2               NaN         0             0            0        0   
3               NaN         0             0            0        0   
4               NaN         0             0            0        0   

   SecurityDelay  LateAircraftDelay  
0              0                  0  
1              0                  0  
2              0                  0  
3              0                  0  
4              0                  0  

[5 rows x 29 columns]

In [11]:
Y05_06.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8595797 entries, 0 to 8595796
Data columns (total 29 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Year               int64  
 1   Month              int64  
 2   DayofMonth         int64  
 3   DayOfWeek          int64  
 4   DepTime            float64
 5   CRSDepTime         int64  
 6   ArrTime            float64
 7   CRSArrTime         int64  
 8   UniqueCarrier      object 
 9   FlightNum          int64  
 10  TailNum            object 
 11  ActualElapsedTime  float64
 12  CRSElapsedTime     float64
 13  AirTime            float64
 14  ArrDelay           float64
 15  DepDelay           float64
 16  Origin             object 
 17  Dest               object 
 18  Distance           int64  
 19  TaxiIn             int64  
 20  TaxiOut            int64  
 21  Cancelled          int64  
 22  CancellationCode   object 
 23  Diverted           int64  
 24  CarrierDelay       int64  
 25  WeatherDelay      

In [13]:
Y05_06.to_sql('Y05_06', con = conn, index = False)

## Overview of Data

### Checking sample size in each year

In [7]:
dfA = pd.read_sql_query('''SELECT Year, COUNT(FlightNum) AS total_flight 
                        FROM 'Y05_06'
                        GROUP BY Year''',conn)
dfA

Year  total_flight
0  2005       7140596
1  2006       7141922

Includes all observations. Even those with missing values. 

### Overview of Delayed Flights
Filtering out Cancelled and Diverted flights

In [8]:
dfB = pd.read_sql_query('''SELECT Year, 
                COUNT(FlightNum) AS flights,
                AVG(ArrDelay) AS avg_arr_delay,
                MAX(ArrDelay) AS max_arr_delay,
                MIN(ArrDelay) AS min_arr_delay
                FROM 'Y05_06'
                WHERE Cancelled = 0 AND Diverted = 0
                GROUP BY Year''',conn)
dfB

Year  flights  avg_arr_delay  max_arr_delay  min_arr_delay
0  2005  6992838       7.181343         1925.0         -939.0
1  2006  7003802       8.682840         1779.0         -592.0

In [9]:
C = pd.merge(
    dfA,
    dfB,
    on='Year')
C

Year  total_flight  flights  avg_arr_delay  max_arr_delay  min_arr_delay
0  2005       7140596  6992838       7.181343         1925.0         -939.0
1  2006       7141922  7003802       8.682840         1779.0         -592.0

Saving C as a csv file

In [11]:
C.to_csv('pysetupdata.csv',header = True, index = False)

### Checking if the dataset consist of only the USA

We filter out all the destination, dest, listed in Y05_06. To cross check with the list of airports. This can be done by creating a data frame 'check' and then parsing in our dest and inner join with the airports.

In [6]:
dest = pd.read_sql_query("SELECT Dest From Y05_06", conn)
dest

Dest
0         ORD
1         ORD
2         ORD
3         ORD
4         BOS
...       ...
14282513  EWR
14282514  ATL
14282515  MCO
14282516  SJC
14282517  SLC

[14282518 rows x 1 columns]

In [10]:
airports = pd.read_sql_query("SELECT * From airports",conn)
airports

iata                    airport              city state country  \
0     00M                   Thigpen        Bay Springs    MS     USA   
1     00R       Livingston Municipal        Livingston    TX     USA   
2     00V                Meadow Lake  Colorado Springs    CO     USA   
3     01G               Perry-Warsaw             Perry    NY     USA   
4     01J           Hilliard Airpark          Hilliard    FL     USA   
...   ...                        ...               ...   ...     ...   
3371  ZEF            Elkin Municipal             Elkin    NC     USA   
3372  ZER  Schuylkill Cty/Joe Zerbey        Pottsville    PA     USA   
3373  ZPH      Zephyrhills Municipal       Zephyrhills    FL     USA   
3374  ZUN                 Black Rock              Zuni    NM     USA   
3375  ZZV       Zanesville Municipal        Zanesville    OH     USA   

            lat        long  
0     31.953765  -89.234505  
1     30.685861  -95.017928  
2     38.945749 -104.569893  
3     42.741347  -78.052081  
4     30.688012  -81.905944  
...         ...         ...  
3371  36.280024  -80.786069  
3372  40.706449  -76.373147  
3373  28.228065  -82.155916  
3374  35.083227 -108.791777  
3375  39.944458  -81.892105  

[3376 rows x 7 columns]

In [21]:
check = pd.merge(dest,airports,how = "inner", left_on = "Dest", right_on = "iata")

In [36]:
check

dest iata                       airport state country
0         ORD  ORD  Chicago O'Hare International    IL     USA
1         ORD  ORD  Chicago O'Hare International    IL     USA
2         ORD  ORD  Chicago O'Hare International    IL     USA
3         ORD  ORD  Chicago O'Hare International    IL     USA
4         ORD  ORD  Chicago O'Hare International    IL     USA
...       ...  ...                           ...   ...     ...
14282513  TTN  TTN        Trenton-Mercer County     NJ     USA
14282514  TTN  TTN        Trenton-Mercer County     NJ     USA
14282515  TTN  TTN        Trenton-Mercer County     NJ     USA
14282516  TTN  TTN        Trenton-Mercer County     NJ     USA
14282517  TTN  TTN        Trenton-Mercer County     NJ     USA

[14282518 rows x 5 columns]

In [37]:
check.describe()

dest      iata                            airport     state  \
count   14282518  14282518                           14282518  14273496   
unique       301       301                                301        52   
top          ATL       ATL  William B Hartsfield-Atlanta Intl        CA   
freq      834629    834629                             834629   1685756   

         country  
count   14282518  
unique         1  
top          USA  
freq    14282518

With that we conclude the following results: 
Total Airports: 301
Total states : 52
Total Country : 1

### Disconnecting from DB

In [12]:
conn.close()